pkg/xds/envoy/listeners/fault_injection_configurer.go

```diff
@@ -0,0 +1,120 @@
+package listeners
+
+import (
+	envoy_listener "github.com/envoyproxy/go-control-plane/envoy/api/v2/listener"
+	envoy_api_v2_route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+	envoy_filter_fault "github.com/envoyproxy/go-control-plane/envoy/config/filter/fault/v2"
+	envoy_http_fault "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/fault/v2"
+	envoy_hcm "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	envoy_type_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher"
+	envoy_wellknown "github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/wrappers"
+
+	"github.com/Kong/kuma/api/mesh/v1alpha1"
+	"github.com/Kong/kuma/pkg/core/resources/apis/mesh"
+	"github.com/Kong/kuma/pkg/xds/envoy/routes"
+	"github.com/Kong/kuma/pkg/xds/envoy/tags"
+)
+
+func FaultInjection(faultInjection *mesh.FaultInjectionResource) FilterChainBuilderOpt {
+	return FilterChainBuilderOptFunc(func(config *FilterChainBuilderConfig) {
+		config.Add(&FaultInjectionConfigurer{
+			faultInjection: faultInjection,
+		})
+	})
+}
+
+type FaultInjectionConfigurer struct {
+	faultInjection *mesh.FaultInjectionResource
+}
+
+func (f *FaultInjectionConfigurer) Configure(filterChain *envoy_listener.FilterChain) error {
+	if f.faultInjection == nil {
+		return nil
+	}
+
+	config := &envoy_http_fault.HTTPFault{
+		Delay:   convertDelay(f.faultInjection.Spec.Conf.GetDelay()),
+		Abort:   convertAbort(f.faultInjection.Spec.Conf.GetAbort()),
+		Headers: createHeaders(f.faultInjection.Spec.SourceTags()),
+	}
+
+	rrl, err := convertResponseRateLimit(f.faultInjection.Spec.Conf.GetResponseBandwidth())
+	if err != nil {
+		return err
+	}
+	config.ResponseRateLimit = rrl
+
+	pbst, err := ptypes.MarshalAny(config)
+	if err != nil {
+		return err
+	}
+
+	return UpdateHTTPConnectionManager(filterChain, func(manager *envoy_hcm.HttpConnectionManager) error {
+		manager.HttpFilters = append([]*envoy_hcm.HttpFilter{
+			{
+				Name: envoy_wellknown.Fault,
+				ConfigType: &envoy_hcm.HttpFilter_TypedConfig{
+					TypedConfig: pbst,
+				},
+			},
+		}, manager.HttpFilters...)
+		return nil
+	})
+}
+
+func createHeaders(tagSets []v1alpha1.SingleValueTagSet) (matchers []*envoy_api_v2_route.HeaderMatcher) {
```

Comment:
I might be reading something wrong, but I think there is a problem here. When I specify such policy
```
sources:
- service: payments
- service: ratings
destination:
- redis
conf:
  delay:
    value: 10ms
    percentage: 100.0
```
I want to slow down the connection between payments->redis and ratings->redis.

Multiple matchers in the sources section are OR not AND.
That means that we have to generate Fault Injection filter for every source matcher object

Reply:
Every fault injection filter has multiple headers matchers, which work as OR. So we just map every source matcher on every header matcher. So I guess case you described will work

Reply:
Sorry about that, you are right! Can we extend the `fault_injection_configurer_test.go` so we have this case covered?

One more thing. Are we supporting `*` in regex?
you can specify such policy
```
sources:
- service: *
```
or
```
sources:
- service: *
  cloud: aws
```

Can you please also add test cases for it?

Reply:
Yes, we are supporting `*`. There is already test cases for this: https://github.com/Kong/kuma/pull/649/files#diff-9c5af35e9a515658d55c693807d6d892R69

Reply:
> Can we extend the fault_injection_configurer_test.go so we have this case covered?

Sure, I will add test for it